controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +141,131 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+		if err != nil {
+			return nil, err
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A Cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
+	m, err := r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+	clusterState.machineDeployments = m
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterTopologyReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.InfrastructureRef.Kind, cluster.Spec.InfrastructureRef.Name)
+	}
+	return infra, nil
+}
+
+// getCurrentControlPlaneState returns information on the ControlPlane being used by the Cluster. If a reference is not found,
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterTopologyReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.ControlPlaneRef.Kind, cluster.Spec.ControlPlaneRef.Name)
+	}
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the ClusterClass of the given Cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference %s for ControlPlane, %s", cp.GetKind(), cp.GetName())
```

Comment:
```suggestion
		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference %s, %s", cp.GetKind(), cp.GetName())
```

@killianmuldoon if you want. I intentionally dropped the control plane in my suggestion.

Otherwise it reads something like `failed to get InfrastructureMachineTempalte reference KubeadmControlPlane for ControlPlane, <control-plane-name>` which sounds a bit strange to me. (also it would be then the same error as below which seems to make sense)